compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/TypeChecker.java

```diff
@@ -1481,51 +1497,193 @@ private boolean exprIncompatible(BType eType, BLangExpression expr) {
+
+        if (expType.tag != TypeTags.READONLY) {
+            return tupleType;
+        }
+
+        tupleType.flags |= Flags.READONLY;
+        return tupleType;
     }
 
     public void visit(BLangRecordLiteral recordLiteral) {
         int expTypeTag = expType.tag;
 
-        if (expTypeTag == TypeTags.NONE) {
-            expType = defineInferredRecordType(recordLiteral);
+        if (expTypeTag == TypeTags.NONE || expTypeTag == TypeTags.READONLY) {
+            expType = defineInferredRecordType(recordLiteral, expType);
         } else if (expTypeTag == TypeTags.OBJECT) {
-            dlog.error(recordLiteral.pos,
-                       DiagnosticCode.INVALID_RECORD_LITERAL, expType);
+            dlog.error(recordLiteral.pos, DiagnosticCode.INVALID_RECORD_LITERAL, expType);
             resultType = symTable.semanticError;
             return;
         }
 
-        resultType = checkMappingConstructorCompatibility(expType, recordLiteral);
+        resultType = getEffectiveMappingType(recordLiteral,
+                                             checkMappingConstructorCompatibility(expType, recordLiteral));
+    }
+
+    private BType getEffectiveMappingType(BLangRecordLiteral recordLiteral, BType applicableMappingType) {
+        if (applicableMappingType == symTable.semanticError ||
+                (applicableMappingType.tag == TypeTags.RECORD && Symbols.isFlagOn(applicableMappingType.flags,
+                                                                                  Flags.READONLY))) {
+            return applicableMappingType;
+        }
+
+        Map<String, RecordLiteralNode.RecordField> readOnlyFields = new LinkedHashMap<>();
+        LinkedHashMap<String, BField> applicableTypeFields =
+                applicableMappingType.tag == TypeTags.RECORD ? ((BRecordType) applicableMappingType).fields :
+                        new LinkedHashMap<>();
+
+        for (RecordLiteralNode.RecordField field : recordLiteral.fields) {
+            if (field.getKind() == NodeKind.RECORD_LITERAL_SPREAD_OP) {
+                continue;
+            }
+
+            String name;
+            if (field.isKeyValueField()) {
+                BLangRecordKeyValueField keyValueField = (BLangRecordKeyValueField) field;
+
+                if (!keyValueField.readonly) {
+                    continue;
+                }
+
+                BLangExpression keyExpr = keyValueField.key.expr;
+                if (keyExpr.getKind() == NodeKind.SIMPLE_VARIABLE_REF) {
+                    name = ((BLangSimpleVarRef) keyExpr).variableName.value;
+                } else {
+                    name = (String) ((BLangLiteral) keyExpr).value;
+                }
+            } else {
+                BLangRecordVarNameField varNameField = (BLangRecordVarNameField) field;
+
+                if (!varNameField.readonly) {
+                    continue;
+                }
+                name = varNameField.variableName.value;
+            }
```

Comment:
I think we should consider providing common info such as the name, type of a field through an API. Would simplify code like this. Again, I mean as a future improvement.

Reply:
Yeah, +1.